pkg/controller/shared_state.go

```diff
@@ -1820,3 +1820,122 @@ func (c *SharedState) incrementAttachmentCount(ctx context.Context, node string)
+	desiredAzVolumeAttachment := &azdiskv1beta2.AzVolumeAttachment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: azVolumeAttachmentName,
+			Labels: map[string]string{
+				consts.NodeNameLabel:   nodeName,
+				consts.VolumeNameLabel: volumeName,
+				consts.RoleLabel:       string(role),
+			},
+			// if the volumeAttachment shows not yet attached, and VolumeAttachRequestAnnotation needs to be set from the controllerserver
+			// if the volumeAttachment shows attached but the actual volume isn't due to controller restart, VolumeAttachRequestAnnotation needs to be set by the noderserver during NodeStageVolume
+			Finalizers: []string{consts.AzVolumeAttachmentFinalizer},
+		},
+		Spec: azdiskv1beta2.AzVolumeAttachmentSpec{
+			VolumeName:    volumeName,
+			VolumeID:      volumeHandle,
+			NodeName:      nodeName,
+			RequestedRole: role,
+			VolumeContext: volumeAttributes,
+		},
+	}
+	azureutils.AnnotateAPIVersion(desiredAzVolumeAttachment)
+
+	return desiredAzVolumeAttachment
+}
+
+func (c *SharedState) recreateAzVolumeAttachment(ctx context.Context, azVolumeAttachmentName string, desiredAzVolumeAttachment *azdiskv1beta2.AzVolumeAttachment) error {
+	w, _ := workflow.GetWorkflowFromContext(ctx)
+
+	statusUpdateRequired := true
+	// check if the CRI exists already
+	azVolumeAttachment, err := azureutils.GetAzVolumeAttachment(ctx, c.cachedClient, c.azClient, azVolumeAttachmentName, c.config.ObjectNamespace, false)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			w.Logger().Infof("Recreating AzVolumeAttachment(%s)", azVolumeAttachmentName)
+			azVolumeAttachment, err = c.azClient.DiskV1beta2().AzVolumeAttachments(c.config.ObjectNamespace).Create(ctx, desiredAzVolumeAttachment, metav1.CreateOptions{})
+			if err != nil {
+				w.Logger().Errorf(err, "failed to create AzVolumeAttachment (%s)", azVolumeAttachmentName)
+				return err
+			}
+		} else {
+			w.Logger().Errorf(err, "failed to get AzVolumeAttachment (%s)", azVolumeAttachmentName)
+			return err
+		}
+	} else if apiVersion, ok := azureutils.GetFromMap(azVolumeAttachment.Annotations, consts.APIVersion); !ok || apiVersion != azdiskv1beta2.APIVersion {
+		w.Logger().Infof("Found AzVolumeAttachment (%s) with older api version. Converting to apiVersion(%s)", azVolumeAttachmentName, azdiskv1beta2.APIVersion)
+
+		for k, v := range desiredAzVolumeAttachment.Labels {
+			azVolumeAttachment.Labels = azureutils.AddToMap(azVolumeAttachment.Labels, k, v)
+		}
+
+		for k, v := range azVolumeAttachment.Annotations {
+			azVolumeAttachment.Status.Annotations = azureutils.AddToMap(azVolumeAttachment.Status.Annotations, k, v)
+		}
+
+		// for now, we don't empty the meta annotations after migrating them to status annotation for safety.
+		// note that this will leave some remnant garbage entries in meta annotations
+		for k, v := range desiredAzVolumeAttachment.Annotations {
+			azVolumeAttachment.Annotations = azureutils.AddToMap(azVolumeAttachment.Status.Annotations, k, v)
```

Comment:
Seems like we need to change it to be `azVolumeAttachment.Annotations = azureutils.AddToMap(azVolumeAttachment.Annotations, k, v)`